### PR TITLE
Error handling re-worked

### PR DESCRIPTION
**Pull request by zarnovican**
Hi,

Refer first to #541 for my comments on current issues, if you haven't read it already.

Basically, in this PR I have addressed them as I've seen fit, taking a great liberty of hacking existing code.
#### Summary
- error handler framework implemented which execute custom handlers until a matching handler is found
- handler is installed either by `register_error_handler` method or `error_handler` decorator
- handler may indicate by returning None that processing should continue to next candidate
- first match wins, even if better match was there (subclass is better than base exception class)
- handlers are processed in _reverse_ order they were registered (or decorated). So, user may override any Flask-Restful default handlers as his handler will be registered later, closer to the top.
- existing code in `handle_error` has been re-written in form of error handlers

I didn't get any feedback on these topics, so I decided for myself..

> Much bigger rewrite is needed. It should be clarified first:
> 
> if processing should stop on first matching handler - this will change the existing semantics

It does stop on first error handler which return non-None.

> if existing handle_error code should be rewritten as several smaller error handlers

Yes. Old if-code==XX-then were rewritten as error handlers and they are installed by default.

> what to do when exception is thrown inside error handler (may be explicit abort(404))

Exception is logged and http 500 is returned to client. No further processing is done.
#### Notes
- although all tests are green, I'm not 100% sure that nothing was broken. Actually, I'm aware that in some cases new code returns slightly different response that old one. This is due to the fact that handlers are context-free and they cannot pass data from one to the other.
- I took `self.unauthorized` method and changed its signature to match error handler (and updated all tests referring that method). I'm not aware if `unauthorized` was used somewhere else except 401 handling. If yes, than new error handler should be created and this one left for compatibility.
- I had to update one test to make it pass. Old code was catching all HTTPExceptions with code 401. New code is handling only Unauthorized (subclass of HTTPException with code 401).
- handling of 406 could not be done as error handler (handler cannot affect final representation of data). That code had to be moved to `make_response`
- Bunch of new methods were created. As they are all visible as API, I would be more than happy to rename them to follow any project naming conventions / personal preferences / etc.
- Sphinx documentation should be written. But I won't invest my time before this PR is at least informally approved

### CODE REVIEW

**Comment by danijar** [comment]
Looks quite good. I would like to be able to raise exceptions inside error handlers to escalate responsibility. For example:

```python
    @api.errorhandler(ValueError)
    def handle_value_error(error):
        print('Caught', error.message)
        if error.message == 'No JSON object could be decoded':
            raise BadRequest('JSON content required.')
        raise error
```

This saves a lot of code in the request handlers since common checks can be defined centrally. Here is a [working implementation](https://gist.github.com/danijar/827bee07350a96bbf35b) of that. This also handles inherited errors correctly, i.e. handlers for sub-classes are preferred over handlers for base-class exceptions.

Do you think we could integrate that into your pull request?

**Comment by joshfriend** [comment]
Oh boy, how did I miss this one?

> (from #451) currently it is possible to have both custom and default handler code executed. Not so after the change

I am OK with that.

> first match wins, even if better match was there (subclass is better than base exception class)

I liked @danijar's [implementation](https://github.com/flask-restful/flask-restful/pull/401#issuecomment-189638310) that sorted the error handlers so that the most specific exception handler is found first.

> handler may indicate by returning `None` that processing should continue to next candidate

:+1: If the user wants the handler to have no response body, they can just return an empty `Response` object?

> Exception is logged and http 500 is returned to client. No further processing is done.

:+1:

> handle_error code assumes that data is a dictionary (data.update(custom_data)). This may not always be the case. For example, if you would like to have a custom error handler on non-HTTPException with .data attribute as string. This is a problem on existing code already.

There is no `data` property on Werkzeug exceptions. This is simply a way for a user to put data into an exception for the default handler to return since it was not possible currently to easily override the error handlers to return the data you want. If the user wants to add stuff into an exception that they throw in the app, then they can just override the default handler with something that can handle it.

> I had to update one test to make it pass. Old code was catching all HTTPExceptions with code 401. New code is handling only Unauthorized (subclass of HTTPException with code 401).

:+1: I think your correction is better anyways.

> Sphinx documentation should be written. But I won't invest my time before this PR is at least informally approved

I like this implementation. Please @ mention me when you have new changes or if you need more feedback. I'll try not to loose track of this PR again.

File: flask_restful/__init__.py [hunk 1]
```diff
@@ -287,33 +297,86 @@ def handle_error(self, e):
...
-            }
-            headers = e.get_response().headers
+        for typecheck, handler in self.errorhandlers:
+            if isinstance(e, typecheck):
+                try:
+                    # unpack(None) -> (None, _, _)
+                    data, code, headers = unpack(handler(e))
+                    if data:
+                        break
+                except:
```
Comment (joshfriend) [inline flask_restful/__init__.py]: Never use `except` without an exception class as this will also catch `SystemExit`, `KeyboardInterrupt`, and `GeneratorExit`. It's better to just catch `Exception` instead if you want a generic "catch-all" handler.

File: flask_restful/__init__.py [hunk 2]
```diff
@@ -287,33 +297,86 @@ def handle_error(self, e):
...
+    def default_http_exception_handler(self, e):
+        """Default response for a HTTPException
+
+        :param e: the exception raised while handling the request
+        :type e: werkzeug.exceptions.HTTPException
+        :return: (data, code, headers) tuple
+        """
+        code = e.code
+        default_data = { 'message': getattr(e, 'description', http_status_message(code)), }
+        data = getattr(e, 'data', default_data)
```
Comment (joshfriend) [inline flask_restful/__init__.py]: `HTTPException` has no `data` property as I mentioned in [this comment](https://github.com/flask-restful/flask-restful/pull/544#issuecomment-189654298). I don't think we need to support this anymore and you can just return the `default_data`.

File: flask_restful/__init__.py [hunk 3]
```diff
@@ -287,33 +297,86 @@ def handle_error(self, e):
...
+        :type e: Exception
+        :return: (data, code, headers) tuple or None if no match was found
+        """
+        error_cls_name = type(e).__name__
+        if error_cls_name in self.errors:
+            if isinstance(e, HTTPException):
+                data, code, headers = self.default_http_exception_handler(e)
+            else:
+                data, code, headers = self.default_exception_handler(e)
+            custom_data = self.errors.get(error_cls_name, {})
```
Comment (joshfriend) [inline flask_restful/__init__.py]: With this new errorhandler system, I don't see a reason to keep the old system of initializing the `Api` object with a dict of custom response data mapped to exception names.

**Comment by zarnovican** [comment]
@joshfriend OMG, you guys really are stretching my memory.. I wrote this code 3 months ago !!

> Never use except without an exception class as this will also catch SystemExit, KeyboardInterrupt, and GeneratorExit.

Valid point. Fixed.

> HTTPException has no data propert..

I have removed the code that read `.data` property. Although, I could swear, it was there for some compatibility reason I cannot remember anymore.

> ..I don't see a reason to keep the old system of initializing the Api object with a dict of custom response

I would not dare to make such backward compatibility-breaking change. It's up to you as an owner to decide (and implement). Code update should be easy, but there is also documentation to be updated.

> I liked @danijar's implementation that sorted the error handlers so that the most specific exception handler is found first.

I've been toying with that idea myself. @danijar, no disrespect to your code (I haven't had time to study), but I would suggest to look at Flask, too. They have this functionality for a long time on `master`. It would be just a matter of copy-pasting their code. But then, why the hell two such close projects need to have each their own error handler ?!

**re: rebase** - while trying to rebase my branch, I have noticed that there were further changes to error-handling code that would have to be re-integrated.

I have implemented those two small fixes (without testing, I trust my code coverage ;))

To summarize TODOs:
* update doc (must)
* select best match, rather than first match (should)
* remove old error-handling code and doc update (could)
* rebase & re-integrate (must)

Sorry, but since I wrote this PR, my priorities have shifted. I will have to leave it for someone else to implement.

**Comment by danijar** [comment]
@zarnovican You're right, Flask already has this. Can't we use Flask custom handlers for Flask-RESTful errors completely? I they don't call specialized handlers over general ones, I can contribute to Flask instead.

**Comment by billmccord** [comment]
Workaround:
```
handle_exceptions = app.handle_exception
handle_user_exception = app.handle_user_exception
api = Api(app)
app.handle_user_exception = handle_exceptions
app.handle_user_exception = handle_user_exception
```

**Comment by skirsdeda** [comment]
Fast forward to 2018 and this seems to be still very relevant. @joshfriend If this was brought back to a mergeable state, would you consider merging/releasing it? Or are there any alternatives?

**Comment by TVH7** [comment]
So I'm just going to bump this again. 2019 now, still relevant. @joshfriend

**Comment by zarnovican** [comment]
Closing this seven years old PR. I'm going to delete the fork (and branch) in a few days as well.
